### PR TITLE
chore: process migrated tasks

### PR DESCRIPTION
**Pull request by mrehan27**
part of [MBL-220](https://linear.app/customerio/issue/MBL-220/migration-for-track-background-queue-identify-and-events)

### Changes

- Implemented `MigrationProcessor` in Data Pipelines module
- Added `MigrationAssistant` to help migrating old tasks
- Requested migration using `MigrationAssistant` on SDK initialization
- Added `MigrationSDKComponent` so it is easier to satisfy dependencies in migration module
- Extensions to transform `JSONObject` to `JsonObject` recursively

### CODE REVIEW

File: datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * it to JsonObject for processing.
+ */
+fun JSONObject.toJsonObject(): JsonObject = buildJsonObject {
+    val keys = keys()
+    while (keys.hasNext()) {
+        val key = keys.next()
+        val value = get(key)
+        put(key, value.toSerializableJson())
+    }
+}
```
Comment (Shahroz16) [inline datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt:25]:
```suggestion
fun JSONObject.toJsonObject(): JsonObject = buildJsonObject {
    for (key in keys()) {
        put(key, opt(key).toSerializableJson())
    }
}
```

`get` if you want error to be thrown

File: datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Transforms [JSONArray] to [JsonArray] by leveraging kotlinx.serialization.json.
+ */
+fun JSONArray.toJsonArray(): JsonArray = buildJsonArray {
+    for (i in 0 until length()) {
+        val value = get(i)
+        add(value.toSerializableJson())
+    }
+}
```
Comment (Shahroz16) [inline datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt:35]:
```suggestion
fun JSONArray.toJsonArray(): JsonArray = buildJsonArray {
    for (i in 0 until length()) {
        add(opt(i).toSerializableJson())
    }
}
```

File: datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * The method is recursive and can handle nested JSONObjects and JSONArrays.
+ */
+private fun Any?.toSerializableJson(): JsonElement {
+    return when (this) {
+        null -> JsonNull
+        is JSONObject -> this.toJsonObject()
+        is JSONArray -> this.toJsonArray()
+        else -> this.toJsonElement()
+    }
+}
```
Comment (Shahroz16) [inline datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt:48]:
```suggestion
private fun Any?.toSerializableJson(): JsonElement {
    return when (this) {
        null -> JsonNull
        is JSONObject -> toJsonObject()
        is JSONArray -> toJsonArray()
        else -> toJsonElement()
    }
}
```

Comment (Shahroz16) [reply datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt:48]: if we have custom classes with serializers than `Json.encodeToJsonElement()` might be safer than `toJsonElement()`

File: datapipelines/src/main/kotlin/io/customer/sdk/CustomerIO.kt [hunk 4]
```diff
@@ -48,7 +60,7 @@ class CustomerIO private constructor(
     androidSDKComponent: AndroidSDKComponent,
     override val moduleConfig: DataPipelinesModuleConfig,
     overrideAnalytics: Analytics? = null
-) : CustomerIOModule<DataPipelinesModuleConfig>, DataPipelineInstance() {
+) : CustomerIOModule<DataPipelinesModuleConfig>, DataPipelineInstance(), MigrationProcessor {
```
Comment (Shahroz16) [inline datapipelines/src/main/kotlin/io/customer/sdk/CustomerIO.kt:63]: can we probably create a separate class for it? `CustomerIO` class is already too big 😅

File: tracking-migration/src/main/java/io/customer/tracking/migration/MigrationAssistant.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        migrationSiteId = migrationSiteId
+    )
+    private val sitePreferences: SitePreferenceRepository = migrationSDKComponent.sitePreferences
+    private val queue: Queue = migrationSDKComponent.queue
+
+    /**
+     * Starts the migration process by migrating the existing tracking data to the new
+     * data pipelines implementation using provided migration processor.
+     */
+    fun migrate() {
```
Comment (Shahroz16) [inline tracking-migration/src/main/java/io/customer/tracking/migration/MigrationAssistant.kt:29]: can we call it `init` and avoid a method call? this class will only be initialized if `siteId` exists

**Review by Shahroz16 — APPROVED** [review]
Minor improvements suggestions

File: datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt (hunk 1 above)
Comment (mrehan27) [reply datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt]: Updated

File: datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt (hunk 2 above)
Comment (mrehan27) [reply datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt]: Updated

File: datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt (hunk 3 above)
Comment (mrehan27) [reply datapipelines/src/main/kotlin/io/customer/datapipelines/extensions/JsonExtensions.kt:45]: We are actually using `com.segment.analytics.kotlin.core.utilities.toJsonElement` here which has support for many other types and calling `Json.encodeToJsonElement()` internally where needed. The only two types that we needed and weren't support were `org.json.JSONObject` and `org.json.JSONArray` so I added wrapper to it.

File: datapipelines/src/main/kotlin/io/customer/sdk/CustomerIO.kt (hunk 4 above)
Comment (mrehan27) [reply datapipelines/src/main/kotlin/io/customer/sdk/CustomerIO.kt:63]: I was thinking the same initially, let me try to decouple it and share if I see any challenges in it.

Comment (mrehan27) [reply datapipelines/src/main/kotlin/io/customer/sdk/CustomerIO.kt]: I moved code to `TrackingMigrationProcessor`. Let me know how if it looks good to you now.

File: tracking-migration/src/main/java/io/customer/tracking/migration/MigrationAssistant.kt (hunk 5 above)
Comment (mrehan27) [reply tracking-migration/src/main/java/io/customer/tracking/migration/MigrationAssistant.kt]: I avoid it as it reduces code readability. I added some comments and a helper method to improve readability with code in init block.

**Pull request merged**